File: python/rsa_oaep_pss_example.py
"""RSA-3072 OAEP encryption and RSA-PSS signatures using SHA-256."""

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.asymmetric import padding, rsa


def generate_private_key() -> rsa.RSAPrivateKey:
    return rsa.generate_private_key(public_exponent=65537, key_size=3072)


def oaep() -> padding.OAEP:
    return padding.OAEP(
        mgf=padding.MGF1(hashes.SHA256()),
        algorithm=hashes.SHA256(),
        label=None,
    )


def pss() -> padding.PSS:
    return padding.PSS(
        mgf=padding.MGF1(hashes.SHA256()),
        salt_length=hashes.SHA256().digest_size,
    )


def encrypt(public_key: rsa.RSAPublicKey, plaintext: bytes) -> bytes:
    return public_key.encrypt(plaintext, oaep())


def decrypt(private_key: rsa.RSAPrivateKey, ciphertext: bytes) -> bytes:
    return private_key.decrypt(ciphertext, oaep())


def sign(private_key: rsa.RSAPrivateKey, message: bytes) -> bytes:
    return private_key.sign(message, pss(), hashes.SHA256())


def verify(public_key: rsa.RSAPublicKey, message: bytes, signature: bytes) -> bool:
    try:
        public_key.verify(signature, message, pss(), hashes.SHA256())
        return True
    except InvalidSignature:
        return False


def main() -> None:
    private_key = generate_private_key()
    public_key = private_key.public_key()
    message = b"small secret or content-encryption key"
    assert decrypt(private_key, encrypt(public_key, message)) == message
    signature = sign(private_key, message)
    assert verify(public_key, message, signature)
    assert not verify(public_key, message + b"!", signature)
    print("RsaOaepPssExample encryption and signature OK")


if __name__ == "__main__":
    main()

File: java/src/main/java/org/example/crypto/RsaOaepPssExample.java
package org.example.crypto;

import java.nio.charset.StandardCharsets;
import java.security.*;
import java.security.spec.MGF1ParameterSpec;
import java.security.spec.PSSParameterSpec;
import java.util.Arrays;
import javax.crypto.Cipher;
import javax.crypto.spec.OAEPParameterSpec;
import javax.crypto.spec.PSource;

public final class RsaOaepPssExample {
    private RsaOaepPssExample() {}

    public static KeyPair generateKeyPair() throws Exception {
        KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA");
        generator.initialize(3072, CryptoSupport.RNG);
        return generator.generateKeyPair();
    }

    private static OAEPParameterSpec oaepParameters() {
        return new OAEPParameterSpec("SHA-256", "MGF1", MGF1ParameterSpec.SHA256, PSource.PSpecified.DEFAULT);
    }

    public static byte[] encrypt(PublicKey key, byte[] plaintext) throws Exception {
        Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPPadding");
        cipher.init(Cipher.ENCRYPT_MODE, key, oaepParameters(), CryptoSupport.RNG);
        return cipher.doFinal(plaintext);
    }

    public static byte[] decrypt(PrivateKey key, byte[] ciphertext) throws Exception {
        Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPPadding");
        cipher.init(Cipher.DECRYPT_MODE, key, oaepParameters());
        return cipher.doFinal(ciphertext);
    }

    private static PSSParameterSpec pssParameters() {
        return new PSSParameterSpec("SHA-256", "MGF1", MGF1ParameterSpec.SHA256, 32, 1);
    }

    public static byte[] sign(PrivateKey key, byte[] message) throws Exception {
        Signature signature = Signature.getInstance("RSASSA-PSS");
        signature.setParameter(pssParameters());
        signature.initSign(key, CryptoSupport.RNG);
        signature.update(message);
        return signature.sign();
    }

    public static boolean verify(PublicKey key, byte[] message, byte[] signatureValue) throws Exception {
        Signature signature = Signature.getInstance("RSASSA-PSS");
        signature.setParameter(pssParameters());
        signature.initVerify(key);
        signature.update(message);
        return signature.verify(signatureValue);
    }

    public static void main(String[] args) throws Exception {
        KeyPair keys = generateKeyPair();
        byte[] message = "small secret or content-encryption key".getBytes(StandardCharsets.UTF_8);
        byte[] ciphertext = encrypt(keys.getPublic(), message);
        CryptoSupport.require(Arrays.equals(message, decrypt(keys.getPrivate(), ciphertext)), "RSA-OAEP failed");
        byte[] signature = sign(keys.getPrivate(), message);
        CryptoSupport.require(verify(keys.getPublic(), message, signature), "RSA-PSS failed");
        System.out.println("RsaOaepPssExample encryption and signature OK");
    }
}
